language/stdlib/modules/compare.move

```diff
@@ -0,0 +1,99 @@
+// have the guarantee:
+// `Compare::cmp<T>(t1, t2) == LESS_THEN` iff `Compare::cmp<T>(t2, t1) == GREATER_THEN`
+//
+// Move has bytecodes for polymorphic equality `==` and inequality `!=`. The implementation
+// of `Compare::cmp<T>(t1, t2)` also guarantees that
+// `Compare::cmp<T>(t1, t2) == EQUAL` iff `t1 == t2` and (similarly)
+// `Compare::cmp<T>(t1, t2) != EQUAL` iff `t1 != t2`.
+//
+// We implement `Compare.cmp(t1, t2)` using the `vector<u8>`'s produced by `LCS::to_bytes(t1)` and
+// `LCS::to_bytes(t2)`. We compare the vectors first by length, then byte-by-byte from right to left
+// if the lengths are equal. The length comparison is done in the obvious way; we return
+// `LESS_THAN` if len(lcs(t1)) < len(lcs(t2)) and `GREATER_THAN` if len(lcs(t1)) > len(lcs(t2)).
+// This ensures that for all primitive types `T` with `<` and `>` comparison operators exposed in
+// Move bytecode (`u8`, `u64`, `u128`), we have
+// `Compare::cmp<T>(t1, t2) == LESS_THEN` iff `t1 < t2` and (similarly)
+// `Compare::cmp<T>(t1, t2) == GREATER_THEN` iff `t1 > t2`.
+//
+// For all other types, the order is whatever the LCS encoding of the type and the comparison
+// strategy above gives you. One case where the order might be surprising is the `address` type.
+// Addresses are 16 byte hex values that LCS encodes with the identity function. The right to left,
+// byte-by-byte comparison means that (for example)
+// `Compare::cmp(0x01, 0x10)` == LESS_THAN` (as you'd expect), but
+// `Compare::cmp(0x100, 0x001) == LESS_THAN` (as you probably wouldn't expect). Keep this in mind
+// when writing code that compares addresses.
+//
+// Finally, a note on the constants `EQUAL`, `LESS_THAN`, and `GREATER_THAN`. Move does not have
+// signed integers, so we cannot use the usual 0, -1, 1 convention for this. Instead, we define a
+// new convention using u8 constants:
+// const EQUAL = 0u8
+// const LESS_THAN = 1u8
+// const GREATER_THAN = 2u8
+
+address 0x0:
+module Compare {
+    use 0x0::LCS;
+    use 0x0::Vector;
+
+    // Polymorphic comparison based on the LCS encoding of `v1` and `v2`. The lengths of the
+    // LCS-encoded vector are compared first, then the vectors are compared byte-by-byte from right
+    // to left if the lengths are equal.
+    // Returns either `EQUAL` (0u8), `LESS_THAN` (1u8), or `GREATER_THAN` (2u8).
+    public fun cmp<Value>(v1: &Value, v2: &Value): u8 {
+        cmp_vector_u8(&LCS::to_bytes(v1), &LCS::to_bytes(v2))
+    }
+
+    // Compare `v1` and `v2` using (1) the lengths of the vectors and (2) byte-by-byte comparison
+    // from right to left.
+    // Returns either `EQUAL` (0u8), `LESS_THAN` (1u8), or `GREATER_THAN` (2u8).
+    fun cmp_vector_u8(v1: &vector<u8>, v2: &vector<u8>): u8 {
+        let len1 = Vector::length(v1);
+        let len2 = Vector::length(v2);
+        let len_cmp = cmp_u64(len1, len2);
+        if (len_cmp != 0) {
+            // lengths unequal
+            return len_cmp
+        };
```

Comment:
People don't seem to like it, but worth mentioning you don't need the braces around the `return` here

I personally like `if (len_cmp !=0) return len_cmp;` on one line. 

All stories about people not realizing their indented code did not belong to the if block feel like stories before strong static typing and auto formatters. 
(Granted we don't have an auto formatter currently, so I'd be ok with encouraging people to over use braces until we do 😃)

Reply:
> All stories about people not realizing their indented code did not belong to the if block feel like stories before strong static typing and auto formatters.

I see you anticipated my objection perfectly :). I actually think we should require curly braces everywhere, but that's a discussion for another time... Will change to the idiomatic style for now and we can have the great debate later.